File: 김소희/최단경로.java
import java.io.*;
import java.util.*;

public class Main {

    static class Edge {
        int to;
        int cost;

        Edge(int to, int cost) {
            this.to = to;
            this.cost = cost;
        }
    }

    public static void main(String[] args) throws Exception {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int V = Integer.parseInt(st.nextToken());
        int E = Integer.parseInt(st.nextToken());

        int K = Integer.parseInt(br.readLine());

        List<Edge>[] graph = new ArrayList[V + 1];
        for (int i = 1; i <= V; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int i = 0; i < E; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int cost = Integer.parseInt(st.nextToken());
            graph[a].add(new Edge(b, cost));
        }

        int[] dist = new int[V + 1];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[K] = 0;

        // (비용, 정점)
        PriorityQueue<int[]> pq =
                new PriorityQueue<>(Comparator.comparingInt(o -> o[0]));
        pq.offer(new int[]{0, K});

        while (!pq.isEmpty()) {
            int[] cur = pq.poll();
            int cost = cur[0];
            int node = cur[1];

            // 이미 더 좋은 경로가 있으면 스킵
            if (dist[node] < cost) continue;

            for (Edge e : graph[node]) {
                int next = e.to;
                int newCost = cost + e.cost;

                if (newCost < dist[next]) {
                    dist[next] = newCost;
                    pq.offer(new int[]{newCost, next});
                }
            }
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= V; i++) {
            if (dist[i] == Integer.MAX_VALUE) {
                sb.append("INF\n");
            } else {
                sb.append(dist[i]).append('\n');
            }
        }

        System.out.print(sb);
    }
}

File: 김소희/최단경로.py
import heapq
from collections import defaultdict

# 주어진 시작점에서 다른 모든 정점으로의 최단 경로를 구하는 프로그램을 작성하시오.

graph = defaultdict(list)

V, E = map(int, input().split())
K = int(input())

for _ in range(E):
    a, b, cost = map(int, input().split())
    graph[a].append((cost, b))

costL = [float('inf')] * (V+1)
costL[K] = 0
pq = [(0, K)]

while pq:
    # A에서 n으로 가는 비용
    cost, n = heapq.heappop(pq)    
    
    # 현재 기록이 더 좋다면 넘어가기
    if costL[n] < cost:
        continue

    # 현재가 가장 작은 기록이라면 갱신, 다음 노드 탐색
    costL[n] = cost
    for w, next in graph[n]:
        newCost = cost + w
        if newCost < costL[next]:
            costL[next] = newCost
            heapq.heappush(pq, (costL[next], next))

for v in range(1, V+1):
    if not costL[v] == float('inf'):
        print(costL[v])
    else:
        print("INF")
    
    
